clang/lib/CIR/CodeGen/CIRGenModule.h

```diff
@@ -52,10 +62,33 @@ class CIRGenModule : public CIRGenTypeCache {
   /// A "module" matches a c/cpp source file: containing a list of functions.
   mlir::ModuleOp theModule;
 
+  clang::DiagnosticsEngine &diags;
+
   const clang::TargetInfo &target;
 
 public:
+  mlir::ModuleOp getModule() const { return theModule; }
+
+  /// Helpers to convert Clang's SourceLocation to an MLIR Location.
```

Comment:
This should probably document whether the spelling or expansion location is what's expected?

Reply:
I am not familiar with the complexities of Clang's source locations, so I don't know what to document.  I am open to suggestions about what to put here.

Reply:
I don't know the details about CIR, so it's hard for me to say for certain. Maybe an example will help tease this out? Given:
```
#define ten 10
#define foo(x) (x + ten)
int y = foo(12);
```
you get source locations like: https://godbolt.org/z/oG7fq9b1K Notice how you get `Loc=<<source>:3:9 <Spelling=<source>:2:16>>` or similar where the source location of the final expanded token and the spelling location of where the pre-preprocessed token lives are different -- which location does MLIR expect to work with? It could work with the spelling location (where tokens physically reside in the source file), or it could work with the expansion locations (where the tokens logically reside in the source file after preprocessing), or it could be ambivalent about which form it uses and will happily work with whatever.

Reply:
I ran some tests.  ClangIR (both the incubator and this PR) uses the presumed location from the `SourceLocation` when creating an `mlir::Location`.  It honors `#line` directives and records the location where the macro was expanded.  That is the location we most want to keep around.

I changed the comment to say that the presumed location is what is converted.

Reply:
What we get in this case is:
```
module ...  {
  cir.global  external dsolocal @y = #cir.int<22> : !s32i {alignment = 4 : i64} loc(#loc3)
} loc(#loc)
#loc = loc("a.cpp":0:0)
#loc1 = loc("a.cpp":3:1)
#loc2 = loc("a.cpp":3:9)
#loc3 = loc(fused[#loc1, #loc2])
```

> which location does MLIR expect to work with?

Whatever we tell it should hold. We usually take the presumed (expanded) loc (from `getSourceRange` or begin/end):

```
mlir::Location CIRGenFunction::getLoc(SourceLocation SLoc) {
  // Some AST nodes might contain invalid source locations (e.g.
  // CXXDefaultArgExpr), workaround that to still get something out.
  if (SLoc.isValid()) {
    const SourceManager &SM = getContext().getSourceManager();
    PresumedLoc PLoc = SM.getPresumedLoc(SLoc);
    StringRef Filename = PLoc.getFilename();
    return mlir::FileLineColLoc::get(builder.getStringAttr(Filename),
                                     PLoc.getLine(), PLoc.getColumn());
  } else {
    // Do our best...
    assert(currSrcLoc && "expected to inherit some source location");
    return *currSrcLoc;
  }
}
```
My take is that this is good enough for CIRGen purposes and can/should probably be improved when we start working on emitting debug information (where we can go for non-presumed locs if it makes sense), before that it feels designing prematurely to me.